[3/N] Set correct device to CUDA guards

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #134707
* #134701
* __->__ #134357
* #134345

In `collective()`, `pointToPoint()` and `collectiveCoalesced()`, CUDA guards were created with an unset (default) CUDA device. This is the reason for the IMA facing the NaN checker in issue https://github.com/pytorch/pytorch/issues/134062.

With this fix, `torch.cuda.set_device(device)` is not needed to work around the IMA.

Also refactored a couple places where the guard is created -- preferably we create the guard with a known device, rather than setting the device later.

cc @XilunWu @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

kwen2501 committed:
[3/N] Set correct device to CUDA guards

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2132,7 +2129,9 @@ std::shared_ptr<NCCLComm> ProcessGroupNCCL::getNCCLComm(
 
-  at::cuda::OptionalCUDAGuard gpuGuard;
+  // Get the device index
+  auto deviceIndex = device.index();
```
Nit: do we need the deviceIndex var? Looks like we just pass device to guard directly

Ok probably we are using it somewhere, you moved the decl from below to above.

wconstab reviewed (changes requested):
Oh I forgot to demand tests before I stamp. Please add tests :)

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1103,11 +1103,8 @@ void ProcessGroupNCCL::abortCommsFromMap(
     auto& devName = it.first;
     auto& ncclComm = it.second;
-    at::cuda::OptionalCUDAGuard gpuGuard;
     at::DeviceIndex deviceIndex = getIndexFromDeviceKey(devName);
```
For P2P comms, the deviceIndex could be -1 (invalid), as the keys in the map could be non deviceIndex, but rank to rank numbers.

kwen2501 replied:
Thanks for the information. Let me revert this change here and add the above comments into the code.

kwen2501 committed:
Update on "[3/N] Set correct device to CUDA guards"


In `collective()`, `pointToPoint()` and `collectiveCoalesced()`, CUDA guards were created with an unset (default) CUDA device. This is the reason for the IMA facing the NaN checker in issue https://github.com/pytorch/pytorch/issues/134062.

With this fix, `torch.cuda.set_device(device)` is not needed to work around the IMA.

Also refactored a couple places where the guard is created -- preferably we create the guard with a known device, rather than setting the device later.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[3/N] Set correct device to CUDA guards"


In `collective()`, `pointToPoint()` and `collectiveCoalesced()`, CUDA guards were created with an unset (default) CUDA device. This is the reason for the IMA facing the NaN checker in issue https://github.com/pytorch/pytorch/issues/134062.

With this fix, `torch.cuda.set_device(device)` is not needed to work around the IMA.

Also refactored a couple places where the guard is created -- preferably we create the guard with a known device, rather than setting the device later.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[3/N] Set correct device to CUDA guards"


In `collective()`, `pointToPoint()` and `collectiveCoalesced()`, CUDA guards were created with an unset (default) CUDA device. This is the reason for the IMA facing the NaN checker in issue https://github.com/pytorch/pytorch/issues/134062.

With this fix, `torch.cuda.set_device(device)` is not needed to work around the IMA.

Also refactored a couple places where the guard is created -- preferably we create the guard with a known device, rather than setting the device later.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[3/N] Set correct device to CUDA guards"


In `collective()`, `pointToPoint()` and `collectiveCoalesced()`, CUDA guards were created with an unset (default) CUDA device. This is the reason for the IMA facing the NaN checker in issue https://github.com/pytorch/pytorch/issues/134062.

With this fix, `torch.cuda.set_device(device)` is not needed to work around the IMA.

Also refactored a couple places where the guard is created -- preferably we create the guard with a known device, rather than setting the device later.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
> Oh I forgot to demand tests before I stamp. Please add tests :)

Testing with multiple colls now.

kwen2501 committed:
Update on "[3/N] Set correct device to CUDA guards"


In `collective()`, `pointToPoint()` and `collectiveCoalesced()`, CUDA guards were created with an unset (default) CUDA device. This is the reason for the IMA facing the NaN checker in issue https://github.com/pytorch/pytorch/issues/134062.

With this fix, `torch.cuda.set_device(device)` is not needed to work around the IMA.

Also refactored a couple places where the guard is created -- preferably we create the guard with a known device, rather than setting the device later.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang reviewed (approved):
LGTM now, thanks for the fix.

shuqiangzhang commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -388,6 +389,24 @@ def test_nan_p2p(self):
+    @skip_if_lt_x_gpu(2)
+    def test_nan_check(self):
+        # Not expecting an error, NaN check should not make legit code fail
+        os.environ["TORCH_NCCL_NAN_CHECK"] = "1"
```
Should we also set CUDA_LAUNCH_BLOCKING=1 in this test? as this test failed before only it is enabled

kwen2501 replied:
I think I've seen test failing too when CUDA_LAUNCH_BLOCKING=0 (when the test has two all_reduce's).

kwen2501 commented:
@pytorchbot merge

kwen2501 commented:
@pytorchbot merge

kwen2501 commented:
@pytorchbot merge -f "The revert is unfortunately a miss-target; see [D61938582](https://www.internalfb.com/diff/D61938582); re-landing"

kwen2501 commented:
@pytorchbot merge -f "The revert is unfortunately a miss-target; see [D61938582](https://www.internalfb.com/diff/D61938582); re-landing"

kwen2501 committed:
Update on "[3/N] Set correct device to CUDA guards"


In `collective()`, `pointToPoint()` and `collectiveCoalesced()`, CUDA guards were created with an unset (default) CUDA device. This is the reason for the IMA facing the NaN checker in issue https://github.com/pytorch/pytorch/issues/134062.

With this fix, `torch.cuda.set_device(device)` is not needed to work around the IMA.

Also refactored a couple places where the guard is created -- preferably we create the guard with a known device, rather than setting the device later.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge

kwen2501 committed:
Update on "[3/N] Set correct device to CUDA guards"


In `collective()`, `pointToPoint()` and `collectiveCoalesced()`, CUDA guards were created with an unset (default) CUDA device. This is the reason for the IMA facing the NaN checker in issue https://github.com/pytorch/pytorch/issues/134062.

With this fix, `torch.cuda.set_device(device)` is not needed to work around the IMA.

Also refactored a couple places where the guard is created -- preferably we create the guard with a known device, rather than setting the device later.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]
